### PR TITLE
[5.1] Fix "Specified key was too long" for mysql

### PR DESCRIPTION
**Pull request by gentcys**
Learn from the commit [8744632](https://github.com/rails/rails/commit/8744632fb5649cf26cdcd1518a3554ece95a401b) of rails framework.

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
What about other databases though. Your commit affects them all doesn't it?

**Comment by gentcys** [comment]
@GrahamCampbell Ok, i will add a condition to check the database drive.

File: src/Illuminate/Database/Schema/Blueprint.php [hunk 1]
```diff
@@ -450,6 +451,10 @@ public function char($column, $length = 255)
      */
     public function string($column, $length = 255)
     {
+        if ($this->charset == 'utf8mb4' && $length == 255) {
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Schema/Blueprint.php]: this still doesn't seem right

Comment (gentcys) [inline src/Illuminate/Database/Schema/Blueprint.php]: I'm so sorry. I forgot add the charset parameter of the construct function in the Blueprint class.

Comment (gentcys) [inline src/Illuminate/Database/Schema/Blueprint.php]: @GrahamCampbell Hi, i add the check and the charset parameter of Blueprint's construct. Is that right?

**Comment by alexwhitman** [comment]
This isn't correct.  It's only the index length that can be at most 191 when using utf8mb4, the storage length can still be 255.  The logic should really be:

`if driver is mysql and charset is utf8mb4 and string length > 191 then index length = 191`

The sql generated would then have to apply the length to the index.  But that's only the simple case.  If you have a composite key with multiple string fields then the total index length can be at most 191.

**Comment by gentcys** [comment]
@alexwhitman Thanks, but i  don't know that how to modify the index length. I am new to Laravel and PHP.

**Pull request closed without merging**